### PR TITLE
feat: Indexing Service locator

### PR DESCRIPTION
**Pull request by Peeja**
Part of https://github.com/storacha/project-tracking/issues/140

Depends on https://github.com/storacha/js-indexing-service-client/pull/7

## Outstanding Questions:
* [x] What's the right error to return if `queryClaims` fails?
* [x] Have we verified the claims at this point in the code (that is, the signatures)? Does that happen automatically? If not, what's the right Ucanto way to do that?
* [x] Should/can we use Ucanto's Schema for this instead of Zod? I can't tell if they're meant to solve the same problem or not.
* [x] What's the "context id" in `QueryResult.from()`? It doesn't seem like it ends up being used for anything.
* [x] `assert/location` should be updated to include the optional `space` in its `nb`, but that's in `@web3-storage/content-claims`, which doesn't know about spaces in location claims yet. Is the right thing to do to update it there anyhow, or is there another package which should own it?

Closes https://github.com/storacha/project-tracking/issues/166

### CODE REVIEW

File: src/api.ts [hunk 1]
```diff
@@ -15,9 +15,13 @@ export interface Sliceable {
   range: Range
 }
 
-export type FetchOptions = Partial<Abortable> & Partial<Sliceable>
+export interface SpaceLimited {
+  spaces: DID[]
+}
+
+export type FetchOptions = Partial<Abortable> & Partial<Sliceable> & Partial<SpaceLimited>
```
Comment (alanshaw) [inline src/api.ts:22]: ?
```suggestion
export type FetchOptions = Partial<Abortable & Sliceable & SpaceLimited>
```

**Review by alanshaw — APPROVED** [review]
🚀

File: src/locator/indexing-service/index.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    return {
+      ok: {
+        digest,
+        site: [
+          {
+            location: contentLocationClaim.nb.location.map(
+              (loc) => new URL(loc)
+            ),
+            range: { offset: slice.position[0], length: slice.position[1] },
+            space: contentLocationClaim.nb.space?.did()
```
Comment (fforbeck) [inline src/locator/indexing-service/index.js:94]: **question**: Where do we get the existing delegations? Will we use a different service to fetch them on the Freeway worker?

**Review by fforbeck — APPROVED** [review]

File: src/locator/indexing-service/index.js (hunk 2 above)
Comment (Peeja) [reply src/locator/indexing-service/index.js:94]: The location claims (`assert/location`), or the `space/content/serve/transport/http`s? The former are in the response from the Indexing Service (not sure where or if they exist at rest, or if they're only generated on the fly); the latter aren't stored yet, but will be either in Freeway itself or in a separate service that belongs to it. Setting that up is the next step—it'll probably just be a new DB that Freeway talks to directly.

Comment (fforbeck) [reply src/locator/indexing-service/index.js:94]: I was referring to the `space/content/serve/*` delegation - which will be generated by the w3 client when a new space is created. Got it. Just wanted to double-check if it shouldn't be part of this response.

**Comment by alanshaw** [comment]
> * What's the right error to return if `queryClaims` fails?

Just add `QueryError` to the failure types for the locate call result.

* Have we verified the claims at this point in the code (that is, the signatures)? Does that happen automatically? If not, what's the right Ucanto way to do that?

No, should probably be done in the client though. Can you open an issue?

* Should/can we use Ucanto's Schema for this instead of Zod? I can't tell if they're meant to solve the same problem or not.

Yes, might be less code to bundle overall. It doesn't have the transforms (and probably some other things) that zod has though.

* What's the "context id" in `QueryResult.from()`? It doesn't seem like it ends up being used for anything.

Yeah, I'm not sure why it's there either - perhaps @hannahhoward can shed some light.

* `assert/location` should be updated to include the optional `space` in its `nb`, but that's in `@web3-storage/content-claims`, which doesn't know about spaces in location claims yet. Is the right thing to do to update it there anyhow, or is there another package which should own it?

I'd update it there for now.

**Pull request merged**